File: python/day21.py
def parse(line):
    parts = line.split(" ")
    return int(parts[4])

def game1(start_pos):
    def roll():
        cur = 1
        while True:
            yield cur + (cur + 1) + (cur + 2)
            cur += 3

    scores = [0, 0]
    cur_pos = start_pos[:]
    cur_player = 0
    num_rolls = 0
    roll_gen = roll()
    while scores[0] < 1000 and scores[1] < 1000:
        steps = next(roll_gen)
        num_rolls += 1
        pos = cur_pos[cur_player] + steps
        pos = (pos - 1) % 10 + 1
        cur_pos[cur_player] = pos
        scores[cur_player] += pos
        if scores[cur_player] >= 1000:
            break
        cur_player = 1 - cur_player
    return min(s * num_rolls * 3 for s in scores)

def game2(start_pos):
    mem = {}
    def get_num_winning(cur_pos, scores, cur_player):
        nonlocal mem
        key = (cur_pos, scores, cur_player)
        if key in mem:
            return mem[key]
        if scores[0] >= 21 or scores[1] >= 21:
            if scores[0] > scores[1]:
                return (1, 0)
            else:
                return (0, 1)
        res = [0, 0]
        for i in range(1, 4):
            for j in range (1, 4):
                for k in range (1, 4):
                    steps = i + j + k
                    pos = cur_pos[cur_player] + steps
                    pos = (pos - 1) % 10 + 1
                    wins = get_num_winning((pos, cur_pos[1]),
                                            (scores[0] + pos, scores[1]),
                                            1) if cur_player == 0 else \
                    get_num_winning((cur_pos[0], pos),
                                            (scores[0], scores[1] + pos),
                                            0)
                    res[0] += wins[0]
                    res[1] += wins[1]

        mem[key] = res
        return res
    return get_num_winning(tuple(start_pos), (0, 0), 0)

def solution():
    positions = [parse(x.strip()) for x in open('../data/21.txt').readlines()]
    res1 = game1(positions)
    res2 = max(game2(positions))

    print(f"Answer 1: {res1}\nAnswer 2: {int(res2)}")

File: rust/src/day21.rs
use crate::common;
use std::collections::HashMap;

fn game1(start_pos: &[u64]) -> u64 {
    let mut scores: Vec<u64> = vec![0, 0];
    let mut cur_pos = start_pos.to_owned();
    let mut cur_player = 0;
    let mut num_rolls = 0;
    let mut die_val = 1;
    while scores[0] < 1000 && scores[1] < 1000 {
        let steps = die_val + (die_val + 1) + (die_val + 2);
        die_val += 3;
        num_rolls += 1;
        let mut pos = cur_pos[cur_player] + steps;
        pos = (pos - 1) % 10 + 1;
        cur_pos[cur_player] = pos;
        scores[cur_player] += pos;
        if scores[cur_player] >= 1000 {
            break;
        }
        cur_player = 1 - cur_player;
    }
    scores.iter().map(|s| s * num_rolls * 3).min().unwrap()
}

type Mem = HashMap<((u64, u64), (u64, u64), usize), (u64, u64)>;

fn game2(start_pos: &[u64]) -> Vec<u64> {
    fn get_num_winning(
        cur_pos: &(u64, u64),
        scores: &(u64, u64),
        cur_player: usize,
        mem: &mut Mem,
    ) -> (u64, u64) {
        let key = (*cur_pos, *scores, cur_player);
        if mem.contains_key(&key) {
            return mem[&key];
        }
        if scores.0 >= 21 || scores.1 >= 21 {
            return if scores.0 > scores.1 { (1, 0) } else { (0, 1) };
        }

        let mut res = vec![0, 0];
        for i in 1..=3 {
            for j in 1..=3 {
                for k in 1..=3 {
                    let steps = i + j + k;
                    let mut pos = steps
                        + if cur_player == 1 {
                            cur_pos.1
                        } else {
                            cur_pos.0
                        };
                    pos = (pos - 1) % 10 + 1;
                    let wins = match cur_player {
                        0 => {
                            get_num_winning(&(pos, cur_pos.1), &(scores.0 + pos, scores.1), 1, mem)
                        }
                        1 => {
                            get_num_winning(&(cur_pos.0, pos), &(scores.0, scores.1 + pos), 0, mem)
                        }
                        _ => panic!("Bad player index"),
                    };
                    res[0] += wins.0;
                    res[1] += wins.1;
                }
            }
        }
        let tres = (res[0], res[1]);
        mem.insert(key, tres);
        tres
    }
    let mut mem = HashMap::new();
    let res = get_num_winning(&(start_pos[0], start_pos[1]), &(0, 0), 0, &mut mem);
    vec![res.0, res.1]
}

pub(crate) fn solution() {
    let lines = common::read_lines(&common::data_file(21)).unwrap();
    let start_pos: Vec<u64> = lines
        .iter()
        .map(|line| line.split(' ').collect::<Vec<&str>>()[4].parse().unwrap())
        .collect();
    let res1 = game1(&start_pos);
    let res2 = *game2(&start_pos).iter().max().unwrap();
    println!("Answer 1: {:?}\nAnswer 2: {:?}", res1, res2);
}
